File: examples/canonical-terminal/python-click-blessed/taskflow/interactive/pager.py
from __future__ import annotations

from blessed import Terminal

from taskflow.core.filters import filter_jobs


term = Terminal()


def _color_status(status: str) -> str:
    palette = {
        "done": term.green(status),
        "failed": term.red(status),
        "running": term.yellow(status),
        "pending": term.white(status),
    }
    return palette.get(status, status)


def _render_jobs(jobs: list, index: int) -> None:
    print(term.clear + term.home, end="")
    print(term.bold("TaskFlow Monitor"), "- q=quit, arrows=scroll, f=filter")
    window = jobs[max(0, index - 5): index + 6]
    for offset, job in enumerate(window, start=max(0, index - 5)):
        marker = ">" if offset == index else " "
        print(f"{marker} {job.id} {job.name:<24} {_color_status(job.status.value)}")


def run_pager(jobs: list) -> None:
    if not jobs:
        print("No jobs available.")
        return

    visible = list(jobs)
    index = 0
    with term.cbreak(), term.hidden_cursor():
        while True:
            _render_jobs(visible, index)
            key = term.inkey(timeout=1)
            if not key:
                continue
            if key.lower() == "q":
                print(term.normal, end="")
                return
            if key.name == "KEY_UP":
                index = max(0, index - 1)
            elif key.name == "KEY_DOWN":
                index = min(len(visible) - 1, index + 1)
            elif key.lower() == "f":
                print(term.clear + term.home + "Filter status (pending/running/done/failed, blank=clear): ", end="")
                value = input().strip()
                visible = filter_jobs(jobs, status=value or None)
                if not visible:
                    visible = list(jobs)
                index = 0

File: examples/canonical-terminal/ruby-clamp-tty/lib/taskflow/interactive/pager.rb
# frozen_string_literal: true

require "tty-color"
require "tty-reader"
require_relative "../core/filter"

module Taskflow
  module Interactive
    class Pager
      def initialize(jobs)
        @all_jobs = jobs
        @visible_jobs = jobs
        @index = 0
        @reader = TTY::Reader.new
      end

      def run
        @reader.raw do
          loop do
            render
            case @reader.read_keypress
            when "q"
              break
            when "\e[A"
              @index = [@index - 1, 0].max
            when "\e[B"
              @index = [@index + 1, @visible_jobs.length - 1].min
            when "f"
              filter_by_status
            end
          end
        end
      end

      private

      def render
        puts "\e[H\e[2J"
        puts "TaskFlow Pager - q=quit, up/down=scroll, f=filter"
        window_start = [@index - 5, 0].max
        @visible_jobs[window_start, 11].to_a.each_with_index do |job, offset|
          absolute_index = window_start + offset
          marker = absolute_index == @index ? ">" : " "
          puts "#{marker} #{job.id} #{job.name.ljust(24)} #{color_status(job.status)}"
        end
      end

      def filter_by_status
        print "\nFilter status (pending/running/done/failed, blank=clear): "
        input = STDIN.gets&.strip
        @visible_jobs = if input.nil? || input.empty?
                          @all_jobs
                        else
                          Taskflow::Core::Filter.filter(@all_jobs, status: input)
                        end
        @visible_jobs = @all_jobs if @visible_jobs.empty?
        @index = 0
      end

      def color_status(status)
        return status unless TTY::Color.support?

        case status
        when "done" then "\e[32m#{status}\e[0m"
        when "failed" then "\e[31m#{status}\e[0m"
        when "running" then "\e[33m#{status}\e[0m"
        else "\e[37m#{status}\e[0m"
        end
      end
    end
  end
end
